main.py

```diff
@@ -183,167 +190,168 @@ def compose_message(issues):
-            return
-        if number_of_issues > 10:
-            number_of_issues = 10
-            reply.description = "Let's curb that enthusiasm.. just a little"
-            await channel.send(embed=reply)
-
-        author = message.author
-        if command == "q":
-            author = None
-        elif isinstance(author, discord.Member):
-            author = author.nick or author.name
-        elif isinstance(author, discord.User):
-            author = author.name
-
-        await review_q(channel, number_of_issues, author)
-        return
-    return
+`!fq <N> issue/pr`
+Retrieves N least-recently-updated issue/pr in the feedback queue'''
+    await ctx.channel.send(embed=reply)
 
 
-def read_config(filename):
-    with open(filename, "r") as f:
-        return json.load(f)
+@bot.command(name='rand')
+async def rand_cmd(ctx, number_of_issues: int, noun):
+    command_type="rand"
+    number_of_issues, error_reply, command_config=modify_params(number_of_issues=number_of_issues, noun=noun, command_type=command_type)
+    if error_reply.description:
+        await ctx.channel.send(embed=error_reply)
+    await queue_handler(command_type=command_type, channel=ctx.channel, number_of_issues=number_of_issues, author=None, command_config=command_config)
+
+
+@bot.command(name='review')
+async def review_cmd(ctx, number_of_issues: int):
+    command_type="review"
+    number_of_issues, error_reply, command_config=modify_params(number_of_issues=number_of_issues, noun=None, command_type=command_type)
+    if error_reply.description:
+        await ctx.channel.send(embed=error_reply)
+
+    author = ctx.message.author
+    if isinstance(author, discord.Member):
+        author = author.nick or author.name
+    elif isinstance(author, discord.User):
+        author = author.name
+
+    await queue_handler(command_type="review", channel=ctx.channel, number_of_issues=number_of_issues, author=author, command_config=command_config)
+
+
+@bot.command(name='rq')
+async def rq_cmd(ctx, number_of_issues: int):
+    command_type="review"
+    number_of_issues, error_reply, command_config=modify_params(number_of_issues=number_of_issues, noun=None, command_type=command_type)
+    if error_reply.description:
+        await ctx.channel.send(embed=error_reply)
+
+    await queue_handler(command_type="review", channel=ctx.channel, number_of_issues=number_of_issues, author=None, command_config=command_config)
+
+
+@bot.command(name='feedback')
+async def feedback_cmd(ctx, number_of_issues: int, noun="issue"):
+    command_type="feedback"
+    number_of_issues, error_reply, command_config=modify_params(number_of_issues=number_of_issues, noun=noun, command_type=command_type)
+    if error_reply.description:
+        await ctx.channel.send(embed=error_reply)
+
+    await queue_handler(command_type="feedback", channel=ctx.channel, number_of_issues=number_of_issues, author=None, command_config=command_config)
+
+
+@bot.command(name='fq')
```

Comment:
The inverted splitting of logic (flow goes into one function only to be split again) is present still, though the code is cleaner than before.

I'm writing a prototype of what can be done, not necessarily should be done:
```python
def setup(bot, config):
    for name, functor in config["commands"].items():
        def magic_function(config["cmd_config"][name]):
            async with (): # from queu_handler
                setup(magic_variables)  # aka everything before await queue_handler
                get_issues()  # line 200 or so in queue_handler
                functor(magic_variables)
                send_message() # aka the ending of queue_handler

        bot.command(name=name)(magic_function)
```

This function can be itself made a decorator like `@bot.command` since decorators are just syntactic sugar

Reply:
Some pointers:
- The parameters have to be specified for each function that calls the decorator.
  ```
  @bot.command(name='rand')
  async def rand_cmd(ctx, number_of_issues: int, noun):
  ```
  ```
  @bot.command(name='review')
  async def review_cmd(ctx, number_of_issues: int):
  ```
  ```
  @bot.command(name='what')
  async def what_cmd(ctx):
  ```
  The functions require different arguments which have to be specified manually or through if-else while defining the `bot.command(name=name)`

- `bot.command(name=name)` means we have to duplicate `review` and `rq`, `feedback` and `fq` in the config file or if-else handling. 

- I tried putting 
```
            if author:
                issues = pr_with_pending_review(get_pr_details(issues), author)
                # since async islice doesn't exist
                async for issue in issues:
                    chosen_issues.append(issue)
                    if len(chosen_issues) == number_of_issues:
                        break
            else:
                chosen_issues=issues[:number_of_issues]
```
in separate async function, it didn't work so well (it wasn't updating the `chosen_issues` list, giving out no results. It should update in a `def` fn, but there is something about async that was causing this issue. Still don't know the cause.
Help will be appreciated.

- The `if command_type=="rand":` and `if command_type in ["review", "feedback"]:` still need to be handled for this.

Reply:
Opinion needed:
- Remove `fq` and `rq` options, and use `me` as the noun for author specific review PRs. 

`fq` and `feedback` are same, and `rq` is just a differentiator for author, can instead use a noun, as review can only be done for PRs, thus also eliminating the noun, no-noun problem.

Reply:
Invert the logic. `all` for not filtering, else filtered by author.

Reply:
It ain't working :weary:  
It theoretically should. Will look more into it.
Wish there were preprocessor macros in python.

Reply:
The issue is with the loop variables (name, conf), probably [this](https://stackoverflow.com/a/3431699)

I can access `name` anywhere, but not `conf` since `conf=conf` isn't done. Printing `conf` inside the function declaration gives no result, while outside, it can print.

Reply:
Put things in a class, and don't run the loop in global level, but in a function. That'd resolve the issue (i think)

Reply:
That's solved, forgot to write. Problem was late binding as mentioned [here](https://stackoverflow.com/questions/3431676/creating-functions-in-a-loop/3431699#3431699). 
Solved by just passing `name=name` and `conf=conf` extra params to `magic_function`.

Great solution btw (defining functions in a loop) :heart:

Reply:
The `async for issue in issues` is still troubling, maybe it doesn't work well in `if` clauses

Reply:
Try using `for issue in issues` 🤷‍♂️

Reply:
PS: I'd recommend using `partial function` instead of eager bindings since that'd be explicitly informing the reader about the issue. See this https://docs.python-guide.org/writing/gotchas/#late-binding-closures

Reply:
> Try using for issue in issues man_shrugging

AsyncGenerator, can't

Reply:
> partial function

"Return a new partial object which when called will behave like func called with the positional arguments args and keyword arguments keywords."
We don't want an object, rather a function definition.

This doesn't work:
```
    bot.command(name=name)(magic_function(ctx, int(number_of_issues), noun))

```

Plus anyway `name` and `conf` are required in the function.